src/cargo/util/auth.rs

```diff
@@ -133,52 +158,71 @@ pub fn registry_credential_config(
         }
     }
 
-    let (token, credential_process) = if let Some(name) = &name {
+    let (token, credential_process, secret_key, secret_key_subject) = if let Some(name) = &name {
         log::debug!("found alternative registry name `{name}` for {sid}");
         let RegistryConfig {
             token,
+            secret_key,
+            secret_key_subject,
             credential_process,
             ..
         } = config.get::<RegistryConfig>(&format!("registries.{name}"))?;
         let credential_process =
             credential_process.filter(|_| config.cli_unstable().credential_process);
-        (token, credential_process)
+        let secret_key = secret_key.filter(|_| config.cli_unstable().registry_auth);
+        let secret_key_subject = secret_key_subject.filter(|_| config.cli_unstable().registry_auth);
+        (token, credential_process, secret_key, secret_key_subject)
     } else {
         log::debug!("no registry name found for {sid}");
-        (None, None)
+        (None, None, None, None)
     };
 
     let name = name.as_deref();
-    Ok(match (token, credential_process) {
-        (Some(_), Some(_)) => {
-            return {
-                Err(format_err!(
-                    "both `token` and `credential-process` \
-                    were specified in the config for registry `{name}`.\n\
-                    Only one of these values may be set, remove one or the other to proceed.",
-                    name = name.unwrap()
-                ))
+    let err_both = |token_key: &str, proc_key: &str| {
+        Err(format_err!(
+            "both `{token_key}` and `{proc_key}` \
+            were specified in the config for registry `{name}`.\n\
+             Only one of these values may be set, remove one or the other to proceed.",
+            name = name.unwrap()
+        ))
+    };
+    Ok(
+        match (token, credential_process, secret_key, secret_key_subject) {
```

Comment:
Since this has gotten more complicated, it may be worth deduplicating with the crates.io case above.